src/sparsematrix.jl

```diff
@@ -931,6 +931,35 @@ function Matrix(S::AbstractSparseMatrixCSC{Tv}) where Tv
 end
 Array(S::AbstractSparseMatrixCSC) = Matrix(S)
 
+# Copy the structural nonzeros from S to M
+function _copy_nonzeros_to!(M::Matrix, S::SparseMatrixCSC)
+    colptr = getcolptr(S)
+    rowval = getrowval(S)
+    nzval = getnzval(S)
+    for col in 1:length(colptr)-1
+        for i in colptr[col]:(colptr[col+1]-1)
+            row = rowval[i]
+            val = nzval[i]
+            M[row,col] = val
+        end
+    end
+    return M
+end
+
+function Base.copyto!(M::Matrix{<:Number}, S::SparseMatrixCSC{T}) where T<:Number
+    isempty(S) && return M
+    if size(M) == size(S)
+        fill!(M, zero(T))
+        _copy_nonzeros_to!(M, S)
+    elseif length(M) >= length(S)
+        m_view = view(M, 1:length(S))
+        copyto!(m_view, S)
+        return m_view
```

Comment:
This should return `M`, shouldn't it? And do we have a fast `copyto!` for the call one line above?

Reply:
Yes it should (fixed, but perhaps I should be a test)
No, not yet (I see now that it is actually very slow). One option could be something like 
```julia
m_view = view(M, 1:length(S))
copyto!(m_view, sparsevec(S))
```
and then move the fast implementation of `Vector(SparseVector)` to `copyto!`, which we could define for 
`Union{Vector, SubArray{<:Any,1,<:Vector}, SparseVector`

Reply:
How about computing the linear index from `col` and `rowval`, like `ind = col*m+rowval[i]`, and then write into `M[ind]`? Would be good to avoid allocating an entire sparse array object.

Reply:
Do you mean the `sparsevec` call? 
Are there cases when that doesn't share data with `S`?

Reply:
Thanks for the tips @dkarrasch - now `copyto!(::Array, ::SparseMatrixCSC)` works for `Array` with only one function def. - much nicer. 
I'll do some cleanup (and potential fixes) and add tests when I have some more time.

Do you know if there already is an iterator to loop over the stored linear indices instead of calculating them manually? Would be nice in order to make `copyto!` work for a sparse vector as well?

Reply:
You can use [`LinearIndices`](https://docs.julialang.org/en/v1/base/arrays/#Base.LinearIndices), I think. Something like
```julia
linear_indices = LinearIndices(S)
for ...
    idx = linear_indices[row, col]
    A[idx] = val
```

Reply:
Thanks!
I don't see how to write one method for both `SparseVector` and `SparseMatrixCSC` with that though - but I'm probably missing something. 

I was thinking about an iterator of the type `(ind, val) in indices_nzvals(S)` where `S[ind]==val`, in this case to be used as
```julia
for (ind, val) in indices_nzvals(S)
 A[ind] = val
end
```